homeassistant/components/vizio/media_player.py

```diff
@@ -37,32 +40,47 @@
 MIN_TIME_BETWEEN_FORCED_SCANS = timedelta(seconds=1)
 MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
 
-SUPPORTED_COMMANDS = SUPPORT_TURN_ON | SUPPORT_TURN_OFF \
-                     | SUPPORT_SELECT_SOURCE \
-                     | SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK \
-                     | SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP \
-                     | SUPPORT_VOLUME_SET
+SUPPORTED_COMMANDS = {
+    'soundbar': (SUPPORT_PLAY | SUPPORT_PAUSE
+                 | SUPPORT_SELECT_SOURCE
+                 | SUPPORT_TURN_ON | SUPPORT_TURN_OFF
+                 | SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP
+                 | SUPPORT_VOLUME_SET),
+    'tv': (SUPPORT_PLAY | SUPPORT_PAUSE
+           | SUPPORT_SELECT_SOURCE
+           | SUPPORT_TURN_ON | SUPPORT_TURN_OFF
+           | SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK
+           | SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP
+           | SUPPORT_VOLUME_SET)
+}
+
+MAX_VOLUME = {
+    'soundbar': 31.0,
+    'tv': 100.0
+}
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Required(CONF_HOST): cv.string,
-    vol.Required(CONF_ACCESS_TOKEN): cv.string,
+    vol.Optional(CONF_ACCESS_TOKEN): cv.string,
```

Comment:
The token is not anymore required with the new version of the lib?

Reply:
Here's the rub - it's required for TV's but not required for sound bars. I could not figure out how to make this conditionally required, so if you have any thoughts on how I can do that using voluptuous I am all ears. I did make this clear in the documentation

Reply:
Ok, I think it's fine. Maybe there is a way to do it in voluptuous, or if not, I think it's worth adding a check into the code informing the user if it's being misconfigured.

Reply:
what would you suggest I do on that condition? I can catch on `if device_class == 'tv' && auth_token is None` easily, should I just log an error?

Reply:
Yes, if it's not feasible to do with voluptuous, a misconfiguration should be caught in the code like that. Otherwise there will be users, which may not have read the documentation properly, wondering why it is not working correctly.

Reply:
I added a check to fail the component

Reply:
We can write a custom validator that checks if device class is tv and checks if access token is missin, and if so raises `vol.Invalid`.

See the mysensors component for example of custom validator.

Reply:
@MartinHjelmare I am struggling with this. Here's what I tried to do:
```python
def validate_token(conf):
    if conf[CONF_DEVICE_CLASS] == 'tv' and (conf[CONF_ACCESS_TOKEN] == '' or 
        conf[CONF_ACCESS_TOKEN] is None):
            raise vol.Invalid("If {CONF_DEVICE_CLASS} is 'tv' then "
                              "{CONF_ACCESS_TOKEN} is required. Set "
                              "{CONF_DEVICE_CLASS} to 'soundbar' if "
                              "target device is soundbar without auth")
    return conf

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(vol.All({
    vol.Required(CONF_HOST): cv.string,
    vol.Optional(CONF_ACCESS_TOKEN, default=''): cv.string,
    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
    vol.Optional(CONF_SUPPRESS_WARNING, default=False): cv.boolean,
    vol.Optional(CONF_DEVICE_CLASS, default=DEFAULT_DEVICE_CLASS):
        vol.All(cv.string, vol.Lower, vol.In(['tv', 'soundbar'])),
    vol.Optional(CONF_VOLUME_STEP, default=DEFAULT_VOLUME_STEP):
        vol.All(vol.Coerce(int), vol.Range(min=1, max=10)),
}, validate_token))
```

I get the following error: `AssertionError: Both schemas must be dictionary-based`

I've tried looking at `mysensors` and other components that raise their own `vol.Invalid` but I'm struggling to understand what I am doing wrong, and Google has turned up empty. Can you help me understand how to implement this?

Reply:
This will actually be validated during the validate_setup() step so custom config validation is arguably not necessary (although it would be a nice enhancement if someone can provide some guidance on how to implement it). Log text has been updated to clarify as an alternative solution

Reply:
Just to add in case you have any further thoughts on this, following the guidance here: https://github.com/alecthomas/voluptuous/issues/56

I implemented the following and got the same error:
```python
class VizioSchema(HASchema):
    """Schema class that validates presence of ACCESS_TOKEN when DEVICE_CLASS=tv."""

    def __call__(self, data):
        """Override __call__ to raise validation error if DEVICE_CLASS=tv and ACCESS_TOKEN is None or empty string."""
        config = super().__call__(data)
        token = conf.get(CONF_ACCESS_TOKEN)
        if config[CONF_DEVICE_CLASS] == 'tv' and (token is None or token == ''):
            raise vol.Invalid("When {CONF_DEVICE_CLASS} is 'tv' then "
                              "{CONF_ACCESS_TOKEN} is required.")
        return config


PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    VizioSchema({
        vol.Required(CONF_HOST): cv.string,
        vol.Optional(CONF_ACCESS_TOKEN, default=''): cv.string,
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
        vol.Optional(CONF_SUPPRESS_WARNING, default=False): cv.boolean,
        vol.Optional(CONF_DEVICE_CLASS, default=DEFAULT_DEVICE_CLASS):
            vol.All(cv.string, vol.Lower, vol.In(['tv', 'soundbar'])),
        vol.Optional(CONF_VOLUME_STEP, default=DEFAULT_VOLUME_STEP):
            vol.All(vol.Coerce(int), vol.Range(min=1, max=10)),
    })
)
```

Reply:
Wrap the whole schema with `vol.All`.
```py
PLATFORM_SCHEMA = vol.All(PLATFORM_SCHEMA.extend({...}), validate_token)
```

Reply:
Thank you @MartinHjelmare now it makes perfect sense. Submitting the fix soon